Detectors/TRD/calibration/src/PadCalibCCDBBuilder.cxx

```diff
@@ -0,0 +1,551 @@
+
+  return hDetFilled;
+}
+
+std::vector<std::vector<int>> PadCalibCCDBBuilder::findEmpty(TH2F* hDetectorMap)
+{ // finds the coordinates (col,row) of all empty bins
+  std::vector<std::vector<int>> emptyBins;
+
+  for (int irow = 0; irow < hDetectorMap->GetNbinsY(); irow++) {
+    for (int icolumn = 0; icolumn < hDetectorMap->GetNbinsX(); icolumn++) {
+      float gain = hDetectorMap->GetBinContent(icolumn + 1, irow + 1);
+      if (gain == 0) {
+        std::vector<int> coordinates; // todo fixed length and then set col/row as std::vector[0/1] = xx?
+        coordinates.push_back(icolumn);
+        coordinates.push_back(irow);
+        emptyBins.push_back(coordinates);
+      }
+    } // loop over columns
+  }   // loop over rows
+  return emptyBins;
+}
+
+std::vector<std::vector<int>> PadCalibCCDBBuilder::findInhomogeneities(TH2F* hDet, float allowedDifference)
+{ // finds bins that have 1+ neighbour with significantly different content
+  // gives the coordonates of the pair
+  std::vector<std::vector<int>> suspiciousBinPairs;
+
+  // check for edges - this operation goes before filling
+  // edges along X:
+  int xFirst = hDet->FindFirstBinAbove();
+  int xLast = hDet->FindLastBinAbove(); // stop the process at penultimate vs ultimate row/column
+  // // edges along Y:
+  int yFirst = hDet->FindFirstBinAbove(0, 2);
+  int yLast = hDet->FindLastBinAbove(0, 2); // stop the process at penultimate vs ultimate row/column
+
+  // int nbins_x = hDet->GetNbinsX();
+  // int nbins_y = hDet->GetNbinsY();
+  for (int irow = 0; irow < hDet->GetNbinsY(); irow++) {
+
+    int thisRow = hDet->GetXaxis()->FindBin(irow);
+    // if( thisRow < yFirst || thisRow > yLast ) continue;
+
+    for (int icolumn = 0; icolumn < hDet->GetNbinsX(); icolumn++) {
+
+      int thisColumn = hDet->GetXaxis()->FindBin(icolumn);
+      // if( thisColumn < xFirst || thisColumn > xLast ) continue;
+
+      std::vector<int> pair = compareGain(hDet, thisColumn, thisRow, 1, 0, allowedDifference);
+      if (!(std::any_of(pair.begin(), pair.end(), [](int i) { return i == -1; }))) {
+        suspiciousBinPairs.push_back(pair);
+      }
+      pair.clear();
+
+      pair = compareGain(hDet, thisColumn, thisRow, 0, 1, allowedDifference);
+      if (!(std::any_of(pair.begin(), pair.end(), [](int i) { return i == -1; }))) {
+        suspiciousBinPairs.push_back(pair);
+      }
+
+    } // loop over columns
+
+  } // loop over rows
+
+  return suspiciousBinPairs;
+}
+
+float PadCalibCCDBBuilder::getAverageFromNeighbors(TH2F* hDet, int column, int row, int nbuffer)
+{ // takes bin (column,row) from hDet map
+  // and averages the gain in its (up to) 8 neighbours
+  float average = 0.;
+  std::vector<float> gainNeighbor;
+  int offset = nbuffer / 2;
+
+  for (int irow = 0; irow < nbuffer; irow++) {
+
+    int rowNeighbor = (row - offset) + irow;
+    if (rowNeighbor < 0 || rowNeighbor >= hDet->GetNbinsY())
+      continue; // avoids under and overflow
```

Comment:
Going from your comment here.
If rowNeighbor == 0 this would be the underflow in hDet for any bin and
rowNeigbor >= hDet->GetNBinsY would not only exclude the overflow bin but also the the before it.
Would also apply to colNeighbor
See: https://root.cern/doc/master/classTH1.html#binning
Note: Double check this, might very well be wrong.

Reply:
In this very case, I use row/colNeighbor as hDet->GetXaxis()->FindBin(colNeighbor), so if colNeighbor is 0 it will point to bin 1. Similarly for colNeighbor = GetNbinsX-1, it will call the last bin, i.e., GetNbinsX.
So it should work alright but is a part of another larger issue where I use inconsistently row and column to either refer to bin coordinates or pad coordinates (i.e. as they come from TRD).
As also pointed in another comment, this needs fixing so that the use is non-ambiguous and consistent in all methods.

Reply:
I am not convinced this is true.
If you call ->FindBin(0)  this returns the underflow bin same ->FindBin(GetNbinsX-1) returns the second to last bin.
Maybe I am misunderstanding the doc for it ([https://root.cern/doc/master/classTH1.html#ae0895b66e993b9e9ec728d828b9336fe](FindBin) but I think this may even have unintended side effects (e.g. extending your axis)

> This function tries to extend the axis if the given point belongs to an under-/overflow bin AND if CanExtendAllAxes() is true.

This refers to TH1::GetBin() which list the problem quite clearly I think?

Note: 
TH1::FindBin() internally calls TAxis::FindBin(), so hist->GetXaxis()->FindBin() is the same as hist->FindBin() but nicer wrapped.

Reply:
No. FindBin(0) returns the bin which holds the value 0. In this case it gives you the first bin, i.e. 1. It is GetBin(0) which gets you the underflow.